### PR TITLE
[Test] Make test_sparse_csr.py device-agnostic for OOT backends

### PR DESCRIPTION
**Pull request by KarhouTam**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* __->__ #198078

### CODE REVIEW

**Comment by KarhouTam** [comment]
@claude review these changes

File: test/test_sparse_csr.py
Comment (fffrog) [inline test/test_sparse_csr.py]:
```Python
if base.untyped_storage().data_ptr() != other.untyped_storage().data_ptr():
                    return False
```

I think we can remove the conditional judgement, because we just instantiate this test class in cpu cuda and xpu side.

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -4359,7 +4216,167 @@ def get_current_meta():
         result = operation(*args, **dict(meta=meta, out=out))
         self.assertEqual(result, expected)
 
-    @onlyCUDA
+
+class TestSparseCSROnCPU(TestCase):
```
Comment (fffrog) [inline test/test_sparse_csr.py]:
```suggestion
class TestSparseCSRGeneric(TestCase):
```

File: test/test_sparse_csr.py [hunk 2]
```diff
@@ -2678,103 +2668,6 @@ def test_sampled_addmm_errors(self, device, dtype):
         with self.assertRaisesRegex(RuntimeError, r"Expected mat2 to have strided layout"):
             torch.sparse.sampled_addmm(a_sparse, a, a_sparse)
 
-    @onlyCPU
```
Comment (fffrog) [inline test/test_sparse_csr.py:2679]: Why is this not `ACCELERATOR`?

File: test/test_sparse_csr.py [hunk 3]
```diff
@@ -2678,103 +2668,6 @@ def test_sampled_addmm_errors(self, device, dtype):
...
-                self.assertEqual(grad_mat, ref_grad_mat)
-
-        run_test(4, 5, 4, 10, False)
-        run_test(4, 4, 4, 16, True)
-
-    @skipIfTorchDynamo()
-    @onlyCPU
-    @dtypes(torch.float32, torch.float64, torch.bfloat16, torch.float16)
-    @precisionOverride({torch.bfloat16: 0.02, torch.float16: 0.01})
-    def test_sparse_mm_reduce(self, device, dtype):
```
Comment (fffrog) [inline test/test_sparse_csr.py:2720]: Why is this not `ACCELERATOR`?

File: test/test_sparse_csr.py
Comment (fffrog) [inline test/test_sparse_csr.py:3586]: This one should be cuda-specific

File: test/test_sparse_csr.py [hunk 4]
```diff
@@ -3779,53 +3675,9 @@ def kernel_impl(*args, **kwargs):
                 )
                 self.assertEqual(res_tri, res_dense)
 
-    @onlyCUDA
```
Comment (fffrog) [inline test/test_sparse_csr.py:3780]: Accelerator here?

File: test/test_sparse_csr.py [hunk 5]
```diff
@@ -4359,7 +4216,167 @@ def get_current_meta():
         result = operation(*args, **dict(meta=meta, out=out))
         self.assertEqual(result, expected)
 
-    @onlyCUDA
```
Comment (fffrog) [inline test/test_sparse_csr.py:4360]: onlyAccelerator, here?

**Review by fffrog — CHANGES_REQUESTED** [review]

**Comment by fffrog** [comment]
@claude review this PR.

**Review by fffrog — APPROVED** [review]
SGTM

File: test/test_sparse_csr.py [hunk 6]
```diff
@@ -3588,6 +3576,9 @@ def test_sparse_to_sparse_compressed(self, device, dtype, coalesced, layout):
             self.assertEqual(torch.tensor(sp_matrix.indices, dtype=torch.int64), plain_indices_mth(pt_matrix))
             self.assertEqual(torch.tensor(sp_matrix.data), pt_matrix.values())
 
+class TestSparseCSRCudaOnly(TestCase):
```
Comment (fffrog) [inline test/test_sparse_csr.py:3581]: Missing `instantiate_device_type_test` for `TestSparseCSRCudaOnly`.

**Review by fffrog — CHANGES_REQUESTED** [review]
Hey @KarhouTam.

I have scanned all the suggetions from claude quickly and I think we should follow all the suggestions from section `Testing` and section `Code Quality`.

File: test/test_sparse_csr.py [hunk 7]
```diff
@@ -1049,7 +1049,9 @@ def is_view_of(base, other):
                 base.device != other.device
             ):
                 return False
-            if base.device.type in ('cpu', 'cuda', 'xpu'):
+            # Note: only validates storage on native device types
+            # because some accelerators, like XLA, do not expose storage
+            if base.device.type in ('cpu', 'cuda', 'xpu', torch._C._get_privateuse1_backend_name()):
```
Comment (fffrog) [inline test/test_sparse_csr.py]: As for this one, I do not agree with claude, I think we can remove this one safely and other accelerator can skip this tesecases if they do not support it.

File: test/test_sparse_csr.py [hunk 8]
```diff
@@ -1822,18 +1811,6 @@ def ref_block_addmv(c, a, b, alpha, beta):
             c = make_tensor((m * block_size,), dtype=dtype, device=device, noncontiguous=noncontiguous)
             self.run_test_block_addmm_addmv(torch.addmv, c, a, b, dtype=dtype, device=device, ref=ref_block_addmv)
 
-    @parametrize("matrix_shape", [(3, 3), (5, 7), (11, 9)], name_fn=lambda x: "shape_{}x{}".format(*x))
-    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
-    @onlyCPU
-    def test_addmv(self, device, dtype, matrix_shape):
```
Comment (fffrog) [inline test/test_sparse_csr.py:1827]: Keep this in its current position and add a new decorator named `skipCUDAIfNoSparseGeneric` to it.

File: test/test_sparse_csr.py [hunk 9]
```diff
@@ -2099,39 +2076,6 @@ def maybe_transpose(cond, m):
             m2 = maybe_transpose(t3, torch.randn(50, 25, device=device).to(dtype))
             _test_addmm_addmv(self, torch.addmm, M, m1, m2, transpose_out=t4, layout=layout, mode="all_sparse")
 
-    @onlyCPU
-    @skipCPUIfNoMklSparse
-    @dtypes(*floating_and_complex_types())
-    @sparse_compressed_nonblock_layouts()
-    def test_addmm_dense_result(self, device, dtype, layout):
```
Comment (fffrog) [inline test/test_sparse_csr.py:2105]: Keep this in its current position and add a new decorator to skip CUDA if CUDA do not support this one yet

File: test/test_sparse_csr.py [hunk 10]
```diff
@@ -2678,103 +2622,6 @@ def test_sampled_addmm_errors(self, device, dtype):
         with self.assertRaisesRegex(RuntimeError, r"Expected mat2 to have strided layout"):
             torch.sparse.sampled_addmm(a_sparse, a, a_sparse)
 
-    @onlyCPU
```
Comment (fffrog) [inline test/test_sparse_csr.py:2679]: Keep this in its current position and add a new decorator to skip CUDA

File: test/test_sparse_csr.py [hunk 11]
```diff
@@ -2678,103 +2622,6 @@ def test_sampled_addmm_errors(self, device, dtype):
...
-            ref_mat = mat.clone()
-
-            if train:
-                sparse.requires_grad_()
-                mat.requires_grad_()
-                dense.requires_grad_()
-                ref_mat.requires_grad_()
-
-            ref_out = torch.mm(dense, ref_mat)
-            out = torch.sparse.mm(sparse, mat, 'sum')
```
Comment (fffrog) [inline test/test_sparse_csr.py:2697]: Enable device parameter here.

File: test/test_sparse_csr.py [hunk 12]
```diff
@@ -2678,103 +2622,6 @@ def test_sampled_addmm_errors(self, device, dtype):
...
 
-    @onlyCPU
-    @dtypes(torch.float32, torch.float64, torch.bfloat16, torch.float16)
-    @precisionOverride({torch.bfloat16: 0.01})
-    def test_sparse_mm_reduce_sum(self, device, dtype):
-        def run_test(m, n, k, nnz, train):
-            sparse = self.genSparseCSRTensor((m, k), nnz, dtype=dtype, device=device, index_dtype=torch.int64)
-            dense = sparse.to_dense()
-
-            mat = torch.randn(k, n, dtype=dtype)
```
Comment (fffrog) [inline test/test_sparse_csr.py:2687]: Enable device parameter here.

File: test/test_sparse_csr.py [hunk 13]
```diff
@@ -2678,103 +2622,6 @@ def test_sampled_addmm_errors(self, device, dtype):
...
-                self.assertEqual(grad_mat, ref_grad_mat)
-
-        run_test(4, 5, 4, 10, False)
-        run_test(4, 4, 4, 16, True)
-
-    @skipIfTorchDynamo()
-    @onlyCPU
-    @dtypes(torch.float32, torch.float64, torch.bfloat16, torch.float16)
-    @precisionOverride({torch.bfloat16: 0.02, torch.float16: 0.01})
-    def test_sparse_mm_reduce(self, device, dtype):
```
Comment (fffrog) [inline test/test_sparse_csr.py:2720]: Ditto, In the same way as test_sparse_mm_reduce_sum

File: test/test_sparse_csr.py [hunk 14]
```diff
@@ -4428,12 +4279,169 @@ def get_meta_with_checks(M, K, N, warn_count=0, sparsity=None):
...
+
 
 instantiate_parametrized_tests(TestSparseCompressed)
 instantiate_device_type_tests(TestSparseCompressedDevice, globals(), allow_xpu=True)
 
-instantiate_device_type_tests(TestSparseCSR, globals())
-instantiate_device_type_tests(TestSparseCompressedTritonKernels, globals())
+instantiate_device_type_tests(TestSparseCSR, globals(), allow_xpu=True)
+instantiate_device_type_tests(TestSparseCompressedTritonKernels, globals(), allow_xpu=True)
+instantiate_device_type_tests(TestSparseCSRGeneric, globals(), only_for="cpu")
```
Comment (fffrog) [inline test/test_sparse_csr.py]:
```suggestion
instantiate_parametrized_tests(TestSparseCSRGeneric, globals(), only_for="cpu")
```

File: test/test_sparse_csr.py [hunk 15]
```diff
@@ -4428,12 +4279,169 @@ def get_meta_with_checks(M, K, N, warn_count=0, sparsity=None):
         self.assertEqual(get_meta_with_checks(32, 32, 64, warn_count=1),
                          dict(GROUP_SIZE_ROW=4, SPLIT_N=4, num_stages=1, num_warps=4))
 
+class TestSparseCSRGeneric(TestCase):
```
Comment (fffrog) [inline test/test_sparse_csr.py:4425]: Remove all the testcases in this test class except for test_csr_nnz

File: test/test_sparse_csr.py [hunk 16]
```diff
@@ -21,8 +21,8 @@
     dtypesIfCUDA,
     instantiate_device_type_tests,
     largeTensorTest,
+    onlyAccelerator,
     onlyCPU,
```
Comment (fffrog) [inline test/test_sparse_csr.py]: Remove this one please

**Review by fffrog — CHANGES_REQUESTED** [review]

**Comment by fffrog** [comment]
@claude review this PR, and check the comments I left just now.

File: test/test_sparse_csr.py [hunk 17]
```diff
@@ -2676,15 +2666,15 @@ def test_sampled_addmm_errors(self, device, dtype):
         with self.assertRaisesRegex(RuntimeError, r"Expected mat2 to have strided layout"):
             torch.sparse.sampled_addmm(a_sparse, a, a_sparse)
 
-    @onlyCPU
     @dtypes(torch.float32, torch.float64, torch.bfloat16, torch.float16)
     @precisionOverride({torch.bfloat16: 0.01})
+    @skipCUDAIfNoSparseGeneric
```
Comment (fffrog) [inline test/test_sparse_csr.py]: I think we need use @skipCUDAIf here, because `_sparse_mm_reduce_impl` only support `SparseCsrCPU` in aten/src/ATen/native/native_functions.yaml

File: test/test_sparse_csr.py [hunk 18]
```diff
@@ -2098,10 +2086,10 @@ def maybe_transpose(cond, m):
             m2 = maybe_transpose(t3, torch.randn(50, 25, device=device).to(dtype))
             _test_addmm_addmv(self, torch.addmm, M, m1, m2, transpose_out=t4, layout=layout, mode="all_sparse")
 
-    @onlyCPU
     @skipCPUIfNoMklSparse
     @dtypes(*floating_and_complex_types())
     @sparse_compressed_nonblock_layouts()
+    @skipCUDAIfNoSparseGeneric
```
Comment (fffrog) [inline test/test_sparse_csr.py]: Using skipCUDAIf here

File: test/test_sparse_csr.py
Comment (fffrog) [inline test/test_sparse_csr.py:2688]: enable device parameter here

File: test/test_sparse_csr.py [hunk 19]
```diff
@@ -2714,13 +2704,13 @@ def run_test(m, n, k, nnz, train):
         run_test(4, 4, 4, 16, True)
 
     @skipIfTorchDynamo()
-    @onlyCPU
     @dtypes(torch.float32, torch.float64, torch.bfloat16, torch.float16)
     @precisionOverride({torch.bfloat16: 0.02, torch.float16: 0.01})
+    @skipCUDAIfNoSparseGeneric
```
Comment (fffrog) [inline test/test_sparse_csr.py]: Using skipCUDAIf here

File: test/test_sparse_csr.py
Comment (fffrog) [inline test/test_sparse_csr.py:3630]: This testcases should be designed for CUDA only, because the op is only registered for SparseCsrCUDA

Comment (fffrog) [inline test/test_sparse_csr.py]: Accelerator-agnostic please

**Review by fffrog — CHANGES_REQUESTED** [review]

**Review by fffrog — APPROVED** [review]
SGTM

File: test/test_sparse_csr.py [hunk 20]
```diff
@@ -2714,13 +2705,13 @@ def run_test(m, n, k, nnz, train):
         run_test(4, 4, 4, 16, True)
 
     @skipIfTorchDynamo()
-    @onlyCPU
     @dtypes(torch.float32, torch.float64, torch.bfloat16, torch.float16)
     @precisionOverride({torch.bfloat16: 0.02, torch.float16: 0.01})
+    @skipCUDAIf(True, "_sparse_mm_reduce_impl only support SparseCsrCPU")
```
Comment (guangyey) [inline test/test_sparse_csr.py]: Hi @KarhouTam, `test_sparse_mm_reduce` was originally tested only on CPU, but the current code change seems to extend its test scope to other devices. Could you please take another look? To keep the CI green, I will revert this PR temporarily. Sorry for the trouble, and you can re-land it later!

**Comment by guangyey** [comment]
@KarhouTam  I added the label  `ciflow/xpu` to enable XPU CI. Please feel free to ping me if anything I can help.

File: test/test_sparse_csr.py (hunk 20 above)
Comment (KarhouTam) [reply test/test_sparse_csr.py]: Got it. Will review it.

Comment (guangyey) [reply test/test_sparse_csr.py]: Thanks for your help~

File: test/test_sparse_csr.py [hunk 21]
```diff
@@ -4426,12 +4421,28 @@ def get_meta_with_checks(M, K, N, warn_count=0, sparsity=None):
...
+            a = torch.sparse_csr_tensor(crow_indices, col_indices, values, (rows, cols))
+            self.assertEqual(a._nnz(), nnz)
+
 
 instantiate_parametrized_tests(TestSparseCompressed)
 instantiate_device_type_tests(TestSparseCompressedDevice, globals(), allow_xpu=True)
 
-instantiate_device_type_tests(TestSparseCSR, globals())
-instantiate_device_type_tests(TestSparseCompressedTritonKernels, globals())
+instantiate_device_type_tests(TestSparseCSR, globals(), allow_xpu=True)
```
Comment (fffrog) [inline test/test_sparse_csr.py]: Hi @KarhouTam 

Please disable `xpu` for TestSparseCSR and TestSparseCompressedTritonKernels

**Review by fffrog — APPROVED** [review]

File: test/test_sparse_csr.py (hunk 20 above)
Comment (KarhouTam) [reply test/test_sparse_csr.py]: As @fffrog commented below, I removed `allow_xpu=True` for classes `TestSparseCSR` and `TestSparseCompressedTritonKernels`. Sorry for the inconvenience.

File: test/test_sparse_csr.py [hunk 22]
```diff
@@ -56,7 +57,6 @@
     load_tests,
     parametrize,
     run_tests,
-    serialTest,
```
Comment (fffrog) [inline test/test_sparse_csr.py]: Restore this one.

File: test/test_sparse_csr.py [hunk 23]
```diff
@@ -241,13 +241,7 @@ def test_layout(self, layout):
...
         self.assertEqual(type(layout), torch.layout)
 
-    # arange(2**31 + 1, int64) is 16GiB and `// rows` holds a second one while it
-    # computes, so the peak is 32GiB, not the 30GB previously declared. Serial
-    # because largeTensorTest can only see the cgroup as a whole: the xdist
-    # workers share one memory limit, so a sibling's allocation can swallow the
-    # headroom this test was just told it had.
-    @serialTest()
-    @largeTensorTest("32GB", "cpu")
+    @largeTensorTest("30GB", "cpu")
```
Comment (fffrog) [inline test/test_sparse_csr.py]: Restore this one.

**Review by fffrog — CHANGES_REQUESTED** [review]

**Review by fffrog — APPROVED** [review]
